carbonserver/carbonserver/api/infra/repositories/repository_runs.py

```diff
@@ -1,59 +1,146 @@
-            raise DBException(
-                error=DBError(code=DBErrorEnum.DATA_ERROR, message=e.orig.args[0])
+        with self.session_factory() as session:
+            db_run = SqlModelRun(
+                id=uuid.uuid4(),
+                timestamp=run.timestamp,
+                experiment_id=run.experiment_id,
             )
+            session.add(db_run)
+            session.commit()
+            session.refresh(db_run)
+            return self.map_sql_to_schema(db_run)
+
+        # db_run = sql_models.Run(
+        #     timestamp=run.timestamp,
+        #     experiment_id=run.experiment_id,
+        # )
+        # try:
+        #     self.db.add(db_run)
+        #     self.db.commit()
+        #     self.db.refresh(db_run)
+        #     return db_run
+        # except exc.IntegrityError as e:
+        #     # Sample error : sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) insert or update on table "emissions" violates foreign key constraint "fk_emissions_runs"
+        #     self.db.rollback()
+        #     raise DBException(
+        #         error=DBError(code=DBErrorEnum.INTEGRITY_ERROR, message=e.orig.args[0])
+        #     )
+        # except exc.DataError as e:
+        #     self.db.rollback()
+        #     # Sample error :  sqlalchemy.exc.DataError: (psycopg2.errors.InvalidTextRepresentation) invalid input syntax for type uuid: "5050f55-406d-495d-830e-4fd12c656bd1"
+        #     raise DBException(
+
+    def get_one_run(self, run_id):
+        """Find the run in database and return it
+
+        :run_id: The id of the run to retreive.
+        :returns: An Run in pyDantic BaseModel format.
+        :rtype: schemas.Run
+        """
+        with self.session_factory() as session:
+            e = session.query(SqlModelRun).filter(SqlModelRun.id == run_id).first()
+            if e is None:
+                return None
+            else:
+                return self.map_sql_to_schema(e)
+
+    def list_runs(self):
+        with self.session_factory() as session:
+            e = session.query(SqlModelRun)
+            if e is None:
+                return None
+            else:
+                runs: List[Run] = []
+                for run in e:
+                    runs.append(self.map_sql_to_schema(run))
+                return runs
 
     @staticmethod
-    def get_db_to_class(self, run: sql_models.Run) -> schemas.Run:
+    def map_sql_to_schema(run: SqlModelRun) -> Run:
         """Convert a models.Run to a schemas.Run
 
         :emission: An Run in SQLAlchemy format.
         :returns: An Run in pyDantic BaseModel format.
         :rtype: schemas.Run
         """
-        return schemas.Run(
+        return Run(
             id=run.id,
             timestamp=run.timestamp,
             emission_id=run.emission_id,
             experiment_id=run.experiment_id,
         )
+
+
+class InMemoryRepository(Runs):
```

Comment:
The InMemory Repo can be removed, as it will be hard to maintain now that we are making the logic evolve.
The tests on the router / service are now enough to validate that the domain interface / logic is respected by the repository.

Reply:
OK, thanks